server/src/main/java/org/opensearch/action/admin/cluster/reroute/TransportClusterRerouteAction.java

```diff
@@ -241,6 +241,11 @@ static class ClusterRerouteResponseAckedClusterStateUpdateTask extends AckedClus
             this.allocationService = allocationService;
         }
 
+        @Override
+        public String getClusterManagerThrottlingKey() {
+            return "cluster-reroute-api";
+        }
+
```

Comment:
What prevents two actions from accidentally using the same string?

Reply:
There is a centralised place i.e. ClusterManagerThrottler. But, there is disconnect between the key stored in each executor vs ClusterManagerThrottler class. We can define constant in throttler class and use them in executor class. But the challenge would be to ensure key is unique per executor type and not executor object as lot of executors are defined inline and doesn't have a class.

Reply:
These all tasks have 1:1 mapping with corresponding `TransportClusterManagerNodeActions`. May be we can leverage this and configure throttling keys against `TransportClusterManagerNodeActions` which is responsible for the task. While initializing the task executor we can get the task type from this map. We can have validation against this map for the duplicate values. 

This way we can also restrict the throttling to the actions which are following `TransportClusterManagerNodeActions` route for submitting tasks to master and have the retry mechanism. 

Thoughts?

I have made sample change for this for few of the tasks for ("put-mapping/create-index/delete-index/update-settings"). Will make necessary changes for other tasks based on the early feedback on approach.

Reply:
@dhwanilpatel :  The question I have is 
1. can we ensure there is a transport action associated with all tasks. 
2. what about the inline cluster state submit tasks, could there be multiple tasks such created from same transport class? 

Also we can register only core tasks in this manner. Have we also thought about how to provide register option to plugins in future?

Reply:
> 1. can we ensure there is a transport action associated with all tasks.

Not with all tasks, but if task wants to use throttling framework they will need to have associated transport cluster manager action as it performs the retry on throttling.

> 2. what about the inline cluster state submit tasks, could there be multiple tasks such created from same transport class?

For those task for which I had configured in this PR, they all submit single tasks only. But you have a good point, as in future some transport action may require to submit multiple tasks. 


I have given more thought and proposing some change in solution.

* We can introduce static method `registerTaskForThrottling(String throttlingKey, boolean retryable)`. 
  * This will have logic for detecting duplicate throttling keys as well.
  * Also through this, throttling framework will get information about if for given throttling key if data node is performing retries or not. Many of the internal task types doesn't extend TransportClusterManagerNodeAction, so they might not have default retry capability on throttling. Based on this flag throttler will not throw throttling exception for them.
* We will call this method from constructor of the service/class which is going to submit tasks, and have them register for all the tasks they are submitting along with confirmation on if retry mechanism is there or not. Something similar to the way we register for various transport actions.


> Also we can register only core tasks in this manner. Have we also thought about how to provide register option to plugins in future?

As per my understanding, plugins can access internal class (MasterService/ClusterService/etc) of Opensearch. They can call this new static method and register for tasks. 
I have less expertise in plugin area, please correct me if my above understanding is not correct.

Reply:
@shwetathareja I have made sample change for reference for above approach as well, please take a look at [[6c67dc9](https://github.com/opensearch-project/OpenSearch/pull/4542/commits/6c67dc9abc71ae2f17eba976a4d79c95634d769a)]

Reply:
@dhwanilpatel : thanks for raising a draft PR. The proposal of registering with ```ClusterManagerTaskThrottler``` looks good. The one thing I am wondering instead of making the register method as static and exposing throttler class out of ClusterManager package. One option is to give method in ClusterService for registering as it is accessible from everywhere and then in turn call MasterService -> Throttler class.

Reply:
Yes we can do that, as you mentioned we will need to do it via ClusterService -> MasterService -> Throttler class. 

Since all other services submits the master tasks via cluster services, they will have access of cluster service for registering key as well. 

Will make change for it.